spark3/src/main/java/org/apache/iceberg/spark/MigrateAction.java

```diff
@@ -0,0 +1,272 @@
+        TableProperties.WRITE_NEW_DATA_LOCATION, tableLocation,
+        "migrated", "true");
+  }
+
+  private static SparkSessionCatalog getSparkSessionCatalogOrFail(
+      SparkSession spark,
+      Identifier destTableName) {
+    CatalogPlugin sessionCat;
+    String[] destNamespace = destTableName.namespace();
+    if (destNamespace != null && destNamespace.length > 0 &&
+        spark.sessionState().catalogManager().isCatalogRegistered(destNamespace[0])) {
+      sessionCat = spark.sessionState().catalogManager().catalog(destTableName.namespace()[0]);
+    } else {
+      sessionCat = spark.sessionState().catalogManager().catalog(CatalogManager.SESSION_CATALOG_NAME());
+    }
+    if (!(sessionCat instanceof SparkSessionCatalog)) {
+      throw new IllegalArgumentException(String.format("Cannot use non Iceberg Catalog. Catalog %s was of class" +
+          " %s but %s is required", sessionCat.name(), sessionCat.getClass(), SparkSessionCatalog.class.getName()));
+    }
+    return (SparkSessionCatalog) sessionCat;
+  }
+
+  @Override
+  public Long execute() {
+    Map<String, String> newTableProperties = new ImmutableMap.Builder<String, String>()
+        .put(TableCatalog.PROP_PROVIDER, "iceberg")
+        .putAll(JavaConverters.mapAsJavaMapConverter(sourceTable.properties()).asJava())
+        .putAll(extraIcebergTableProps(destDataLocation, destMetadataLocation))
+        .putAll(additionalProperties)
+        .build();
+
+    StagedTable stagedTable;
+    try {
+      if (sessionCatalogReplacement) {
+        /*
+         * Spark Session Catalog cannot stage a replacement of a Session table with an Iceberg Table.
+         * To workaround this we create a replcamement table which is renamed after it
+         * is successfully constructed.
+         */
+        stagedTable = destCatalog.stageCreateOrReplace(Identifier.of(
+            destTableName.namespace(),
+            destTableName.name() + REPLACEMENT_NAME),
+            sourceTable.schema(),
+            Spark3Util.toTransforms(sourcePartitionSpec), newTableProperties);
+      } else {
+        stagedTable = destCatalog.stageCreateOrReplace(destTableName, sourceTable.schema(),
+            Spark3Util.toTransforms(sourcePartitionSpec), newTableProperties);
+      }
+    } catch (NoSuchNamespaceException e) {
+      throw new IllegalArgumentException("Cannot create a new table in a namespace which does not exist", e);
+    }
+
+    String stagingLocation = destMetadataLocation;
+    Table icebergTable = ((SparkTable) stagedTable).table();
+
+    LOG.info("Beginning migration of {} to {}", sourceTableName, destTableName);
+    SparkTableUtil.importSparkTable(spark, Spark3Util.toTableIdentifier(sourceTableName), icebergTable,
+        stagingLocation);
+
+    Snapshot snapshot = icebergTable.currentSnapshot();
+    long numMigratedFiles = Long.valueOf(snapshot.summary().get(SnapshotSummary.TOTAL_DATA_FILES_PROP));
+
+    try {
+      stagedTable.commitStagedChanges();
```

Comment:
I think everything after the staged table is created should be in the `try` block so that any exception thrown will roll back any changes that were made.

Reply:
Sounds good, fixed